ios/src/TGMapViewController.mm

```diff
@@ -23,26 +25,116 @@ @interface TGMapViewController ()
 
 @implementation TGMapViewController
 
-- (void)setPosition:(TangramGeoPoint)position {
+- (void)loadSceneFile:(NSString*)path {
+    if (!self.map) {
+        return;
+    }
+
+    self.scenePath = path;
+    self.map->loadScene([path UTF8String]);
+    self.renderRequested = YES;
+}
+
+- (void)loadSceneFileAsync:(NSString*)path {
+    if (!self.map) { return; }
+
+    self.scenePath = path;
+
+    MapReady onReadyCallback = [self, path](void* _userPtr) -> void {
+        if (self.mapViewDelegate && [self.mapViewDelegate respondsToSelector:@selector(mapView:didLoadSceneAsync:)]) {
+            [self.mapViewDelegate mapView:self didLoadSceneAsync:path];
+        }
+
+        self.renderRequested = YES;
+    };
+
+    self.map->loadSceneAsync([path UTF8String], false, onReadyCallback, nullptr);
+}
+
+- (void)queueSceneUpdate:(NSString*)componentPath withValue:(NSString*)value {
+    if (!self.map) { return; }
+
+    self.map->queueSceneUpdate([componentPath UTF8String], [componentPath UTF8String]);
+}
+
+- (void)requestRender {
+    if (!self.map) { return; }
+
+    self.renderRequested = YES;
+}
+
+- (CGPoint)lngLatToScreenPosition:(TGGeoPoint)lngLat {
+    static const CGPoint nullCGPoint = {(CGFloat)NAN, (CGFloat)NAN};
+
+    if (!self.map) { return nullCGPoint; }
+
+    double screenPosition[2];
+    if (self.map->lngLatToScreenPosition(lngLat.longitude, lngLat.latitude,
+        &screenPosition[0], &screenPosition[1])) {
+        return CGPointMake((CGFloat)screenPosition[0], (CGFloat)screenPosition[1]);
+    }
+
+    return nullCGPoint;
+}
+
+- (TGGeoPoint)screenPositionToLngLat:(CGPoint)screenPosition {
+    static const TGGeoPoint nullTangramGeoPoint = {NAN, NAN};
+
+    if (!self.map) { return nullTangramGeoPoint; }
+
+    TGGeoPoint lngLat;
+    if (self.map->screenPositionToLngLat(screenPosition.x, screenPosition.y,
+        &lngLat.longitude, &lngLat.latitude)) {
+        return lngLat;
+    }
+
+    return nullTangramGeoPoint;
+}
+
+- (void)pickFeaturesAt:(CGPoint)screenPosition {
+    if (!self.map && !self.mapViewDelegate) { return; }
+
+    const auto& items = self.map->pickFeaturesAt(screenPosition.x, screenPosition.y);
+
+    if (items.size() == 0) { return; }
+
+    const auto& result = items[0];
+    const auto& properties = result.properties;
+    CGPoint position = CGPointMake(result.position[0], result.position[1]);
+
+    NSMutableDictionary* dictionary = [[NSMutableDictionary alloc] init];
+
+    for (const auto& item : properties->items()) {
+        NSString* key = [NSString stringWithUTF8String:item.key.c_str()];
```

Comment:
Are you guaranteed to have valid items/properties/results the whole way here? If any of these are null/nil at any point you'll get a crash trying to insert into the dictionary here.